core/src/main/java/org/apache/oozie/command/wf/SignalXCommand.java

```diff
@@ -468,21 +468,54 @@ else if (!workflowActionBeanListForForked.isEmpty() && !checkForSuspendNode(work
-        for (WorkflowActionBean workflowActionBean : workflowActionBeanListForForked) {
-            LOG.debug("Starting forked actions parallely : " + workflowActionBean.getId());
-            tasks.add(Services.get().get(CallableQueueService.class).new CallableWrapper<ActionExecutorContext>(
-                    new ForkedActionStartXCommand(wfJob, workflowActionBean.getId(), workflowActionBean.getType()), 0));
-        }
 
         try {
-            List<Future<ActionExecutorContext>> futures = Services.get().get(CallableQueueService.class)
-                    .invokeAll(tasks);
+            /**
+             * The limited thread execution mechanism aims to solve the dead-lock when all active threads are
+             * executing the SignalXCommand's invokeAll method.
+             *
+             * Solution
+             * 1. Need to limit directly invokeAll call when the num of rest threads is less than the tasks
+             * 2. To obtain correct active threads number in callableQueue, the SignalXCommand.class lock is needed.
+             *
+             */
+            CallableQueueService callableQueueService = Services.get().get(CallableQueueService.class);
+            List<Future<ActionExecutorContext>> futures = new ArrayList<>();
+
+            synchronized (SignalXCommand.class) {
+                long limitedRestThreadNum = callableQueueService.getQueueThreadsNumber() - callableQueueService.getThreadActiveCount();
+                if (limitedRestThreadNum < workflowActionBeanListForForked.size()) {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Limited callable queue rest threads number: " + limitedRestThreadNum + ", needed forked task size: "
+                                + workflowActionBeanListForForked.size() + ", tasks will be submitted to queue by async mode.");
+                    }
+                    submitJobByQueuing = true;
+                } else {
+                    if (LOG.isDebugEnabled()) {
+                        LOG.debug("Starting forked actions parallely: " + workflowActionBeanListForForked);
+                    }
+                    for (WorkflowActionBean workflowActionBean : workflowActionBeanListForForked) {
+                        futures.add(
+                                callableQueueService.submit(callableQueueService.new CallableWrapper<ActionExecutorContext>(
+                                        new ForkedActionStartXCommand(wfJob, workflowActionBean.getId(), workflowActionBean.getType()), 0))
+                        );
+                    }
+                }
+            }
+
+            if (!futures.isEmpty()) {
+                long startTime = System.currentTimeMillis();
+                callableQueueService.blockingWait(futures);
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Execution time: " + (System.currentTimeMillis() - startTime) / 1000 + " sec");
+                }
```

Comment:
These lines could be moved up to the end of the above else block then remove "if (!futures.isEmpty()) {"

Reply:
got it